include/ghex/transport_layer/util/barrier.hpp

```diff
@@ -0,0 +1,30 @@
+/*
+ * GridTools
+ *
+ * Copyright (c) 2014-2020, ETH Zurich
+ * All rights reserved.
+ *
+ * Please, refer to the LICENSE file in the root directory.
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+
+namespace gridtools {
+    namespace ghex {
+        namespace tl {
+
+            template <typename TLCommunicator>
+            void barrier(MPI_Comm comm, TLCommunicator& tlcomm)
```

Comment:
@boeschf  Do you mean the MPI communicator?

Reply:
I mean that several threads can enter this function and then call the mpi barrier which may cause problems - actually, I am not sure what the semantics should be in this case.

Reply:
This is defined by MPI. If the communicator is duplicated, multiple threads can call `MPI_Barrier`. Otherwise you get undefined behavior. We don't have to do anything about this - users must provide a correct MPI communicator here.

Reply:
One important thing though is that we have to document that all threads MUST call this barrier. Otherwise UCX will not be progressed correctly. So this would be wrong
```
// UCX comm phase
#pramga omp master
barrier(...);
#pragma omp barrier
```
because some master threads could potentially not reach the barrier and hang on the UCX comm phase, while the peer threads would hang (without progress) on the omp barrier.

Reply:
The call to the barrier is thread-safe, yes. And yes, all threads need to call the barrier. But neither we nor MPI can infer how many threads are running (and whether the number of threads is even constant per rank). This is quite a difficult problem, I think, and would have been easily solved with our previous primitives. Right now, I don't see a good solution.

Reply:
@boeschf I think I see what you mean. We have to keep progressing UCX until all threads have completed the barrier, not each individual thread. That's a problem. We could take an `nthreads` argument here, and keep calling UCX progress untill all threads have completed the MPI barrier. This would have the advantage that threads could pass `MPI_COMM_NULL`, in which case we would not require a special thread communicator. That would also solve the case with different number of threads per rank I think.

Reply:
It seems we would only need an atomic variable to do this.

Reply:
I think the implementation will be more complex since you'll need to single out one thread (let's say thread 0) that does the Ibarrier (while still executing the progress in the wait) and the others just progress. There is no guarantee that all communications have finished one thread 0 completes the Ibarrier, however. It would only guarantee that all threads on one rank have finished communication with their thread 0 and that thread 0 has finished communicating with all threads on all ranks. All other communications may or may not be complete. If we only allow the same number of threads per rank then the problem may be simplified, but not by much. It would then work somehow like you describe above, but I think this would still not be enough as some communications may still be outstanding at this point. If a barrier is called for all possible thread-thread combinations (N-1)*N/2, then this may be enough... not sure though.

Reply:
The procedure would be as follows:
- user calls `barrier` for each thread, passing an MPI communicator or `MPI_COMM_NULL`
- we call `MPI_Ibarrier` for all non-null communicators, and call our ucx progress in the test loop. User has to make sure that the MPI communicator has been constructed in a thread-aware manner (MPI standard)
- once out of the test loop, we increase the atomic counter (also for null communicators)
- then we get to the pure ucx progress loop, which we do until the atomic counter reaches the `nthreads` argument

Threads with NULL communicator would increase the counter immediately and proceed to the ucx progress loop.

We don't have to assure completion of UCX comm, we only have to assure that progress is called. This will assure that we leave the `barrier()` function. Then it's the user's responsibility.

Reply:
MPI_Comm should not be an argument, only tl::comm, from which you get contexts, MPI_comm, etc.
What is missing is the number of threads that participate....